### DISCUSSION TITLE
Technical documentation tooling

### DISCUSSION THREAD

**Discussion by aoudiamoncef** [massalabs/massa#2809] [Documentation]
Actually, we are using reStructuredText files which did the trick but IMHO, it's less featured than other solutions.

I explored different types of technical documentation languages in terme of features and tooling. I found that Asciidoc is the most powerful. I use It even for my personal blog and It's widely used for writing technical books.

I tested : markdown with extensions, Latex*, Asciidoc and reStructuredText. 

Example : We have to include a TOML file and render It in our documentation... I found no way to do this except in Asciidoc. 
```
.Node base configuration 
[source,toml]
----
include::massa$config.toml[]
----
```

Asciidoc is widely used, well documented and GitHub have a native support for rendering Asciidoc files.

[Asciidoc cheatsheet](https://powerman.name/doc/asciidoc.html)

**Comment by qdrn** [comment]
At some point the documentation was on Github with MD/Asciidoc format. The issue is that Github wiki is not really convenient to browse a documentation. We can use another solution than Read The Doc but it's something that takes time. Let's make sure it's really worth it before doing it. What would you suggest ?

**Comment by aoudiamoncef** [comment]
For sure, it'll take time to convert current documentation to Asciidoc, but if it could give us a bunch of features and a friendly syntaxe, so we should think about it.

It's not a big deal to add some attractive HTML/CSS/JS output for Asciidoc files using Jekyll or Hugo. 

We could reuse what I've done to build my personal [blog](https://github.com/aoudiamoncef/maoudia.com) which uses Asciidoc + Hugo

[Jekyll vs Hugo](https://bejamas.io/compare/jekyll-vs-hugo/)

[Jekyll doc themes](https://jekyllthemes.dev/tag/documentation/)

[Hugo doc themes](https://themefisher.com/hugo-documentation-themes)

[Static website generators](https://gist.github.com/briandominick/e5754cc8438dd9503d936ef65fffbb2d)

**Comment by aoudiamoncef** [comment]
@damip It could be interesting to create a new repository called for example `doc.massa.com`. We'll have a complete CI/CD workflow which is not related to Massa main project lifecycle

For hosting part, IMHO CloudFlare pages are the best free option. 

[Github Pages vs CloudFlare pages](https://bejamas.io/compare/github-pages-vs-cloudflare-pages/)

**Comment by qdrn** [comment]
[gitbook](https://github.com/GitbookIO/gitbook) also seems to be accepting Asciidoc. I like it from a UX/UI perspective. If that suits everyone we could use it.

**Reply by aoudiamoncef** [reply]
It's a good solution but it's quite expensive. Please see [pricing](https://docs.gitbook.com/pricing/) section.

**Reply by qdrn** [reply]
There's also a free plan. Do we need the features that are not free ?

**Reply by aoudiamoncef** [reply]
I think that we don't fit the criteria to get [free plan ](https://docs.gitbook.com/pricing/plans/non-profit-and-open-source-discounts)

**Reply by qdrn** [reply]
We could once we are open-source. In any case, I was talking about the https://docs.gitbook.com/pricing/plans#personal-plan.

**Comment by qdrn** [comment]
I agree that it would be nice to be able to directly add files in our documentation. Are there any other benefits to switching the doc ? (I think we could use a [custom domain for RTD](https://docs.readthedocs.io/en/stable/custom-domains.html)).

For me the main priority is adding content to the doc, improving it, etc...

**Reply by aoudiamoncef** [reply]
I tried to render some existing TOMLs files and I didn't find a way to do it. This is the main reason for opening this issue. 

It was a little bit tricky to do because I got an issue with my `asciidoctor` install on Ubuntu, but the result is satisfying : 

![image](https://user-images.githubusercontent.com/22281426/179348897-bf43ed59-858f-4cd6-a7a5-4b56b076ae49.png)

I used Hugo + Learn Hugo theme(i18n) + Asciidoc 

TODO : 

- [ ]  Fix some CSS issues with Asciidoc rendering.
- [x]  Migrate current documentation automatically with [Pandoc](https://pandoc.org/)
- [ ]  fix conversion errors.
- [x]  Render all default configurations files (TOMLs)
- [x] Render different types of diagrams(Mermaid, PlantUML...)
- [x] Add `raw` html, css and js if needed. 
- [x] Complete CI/CD workflow with Github pages + CloudFlare pages.

**Reply by aoudiamoncef** [reply]
@qdrn I'll keep this project in standby in case we need it in the future. Actually, there is others important priorities for Massa team.
